cmd/genallocs/genesis_alloc.go

```diff
@@ -0,0 +1,205 @@
+func GenerateGenesisUnlocks(filename string) ([]GenesisAccount, error) {
+	// Read from allocs file.
+	// Open the JSON file
+	path := filepath.Clean(filename)
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file: %w", err)
+	}
+	defer file.Close()
+
+	allocs, err := decodeGenesisAllocs(file)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate unlocking schedules for each account.
+	for i := range allocs {
+		allocs[i].calculateLockedBalances()
+	}
+	return allocs, nil
+}
+
+// Performs verification tasks on provided unlock info.
+func VerifyGenesisAllocs(filename string, expectedHash common.Hash) ([]GenesisAccount, error) {
+	// Open the JSON file
+	path := filepath.Clean(filename)
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file: %w", err)
+	}
+	defer file.Close()
+
+	hasher := blake3.New(32, nil)
+	// Stream the file to the hasher to avoid holding it all in memory.
+	if _, err := io.Copy(hasher, file); err != nil {
+		fmt.Println("Error hashing file:", err)
+		return nil, err
+	}
+	hash := hasher.Sum(nil)
+	if !bytes.Equal(hash, expectedHash.Bytes()) {
+		return nil, errors.New("invalid genesis unlocks")
+	}
+
+	// Reset file pointer to the beginning before decoding JSON.
+	if _, err := file.Seek(0, io.SeekStart); err != nil {
+		return nil, fmt.Errorf("failed to seek file: %w", err)
+	}
+
+	return decodeGenesisAllocs(file)
+}
+
+func decodeGenesisAllocs(r io.Reader) ([]GenesisAccount, error) {
+	// Decode into slice
+	var accounts []GenesisAccount
+	decoder := json.NewDecoder(r)
+	if err := decoder.Decode(&accounts); err != nil {
+		return nil, fmt.Errorf("failed to decode JSON: %w", err)
+	}
+
+	return accounts, nil
+}
+
+// Will calculate the unlock heights according to the pre-defined unlocking schedule.
+func (account *GenesisAccount) calculateLockedBalances() {
+	account.BalanceSchedule = orderedmap.New[uint64, *big.Int]()
+
+	unlockSchedule := unlockSchedules[account.UnlockSchedule]
+
+	// Calculate total lump sum payment.
+	lumpSumPercentage := new(big.Int).SetUint64(unlockSchedule.lumpSumPercentage)
+	lumpSumAmount := lumpSumPercentage.Mul(account.Award, lumpSumPercentage)
+	lumpSumAmount.Div(lumpSumAmount, common.Big100) // Divide back by 100 to undo percentage.
+
+	// Verify that lumpSum payment is not more than the total vested allocation.
+	if lumpSumAmount.Cmp(account.Vested) > 0 {
+		lumpSumAmount = account.Vested
+	}
+
+	var lumpSumIndex uint64
+	if account.UnlockSchedule == 1 {
+		// If the unlockSchedule is 1, then take the lumpSumMonth from the account definition.
+		lumpSumIndex = account.LumpSumMonth
+	} else {
+		// Otherwise, take it from the schedule definition.
+		lumpSumIndex = unlockSchedule.lumpSumMonth
+	}
+	var lumpSumBalance *big.Int
+	lumpSumBlockNum := lumpSumIndex * params.BlocksPerMonth
+	if currentBalance, ok := account.BalanceSchedule.Get(lumpSumBlockNum); ok {
```

Comment:
i dont think we need this as every account is unique in the list of genesis accounts